File: backend/models/Coupon.js
import mongoose from "mongoose";

const CouponSchema = new mongoose.Schema(
  {
    owner: { type: mongoose.Schema.Types.ObjectId, ref: "User", required: true },

    store: { type: String },
    code: { type: String },
    category: { type: String },
    discount: { type: String },
    expiry: { type: Date,
      required: false,
     },

    rawText: { type: String },     // full OCR text
    image: { type: String },       // base64 or image URL

    giftedFrom: { type: mongoose.Schema.Types.ObjectId, ref: "User", default: null },
    isGifted: { type: Boolean, default: false },
    giftedTo: { type: mongoose.Schema.Types.ObjectId, ref: "User" },

    isTradable: { type: Boolean, default: false },
    tradeNotes: { type: String, default: "" }
  },
  { timestamps: true }
);

export default mongoose.model("Coupon", CouponSchema);

File: backend/src/main/java/com/vaulto/model/Coupon.java
package com.vaulto.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;
import org.springframework.data.mongodb.core.mapping.FieldType;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.time.Instant;
import java.util.Date;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "coupons")
public class Coupon {

    @Id
    @JsonProperty("_id")
    private String id;
    
    @Field(targetType = FieldType.OBJECT_ID)
    private String owner;
    
    private String store;
    private String code;
    private String category;
    private String discount;
    private Date expiry;
    private String rawText;
    private String image;
    
    @Field(targetType = FieldType.OBJECT_ID)
    private String giftedFrom;
    
    private boolean isGifted;
    
    @Field(targetType = FieldType.OBJECT_ID)
    private String giftedTo;
    
    private boolean isTradable;
    private String tradeNotes;
    
    private Integer usageCount;
    private Date lastUsedAt;
    
    @CreatedDate
    private Instant createdAt;
    
    @LastModifiedDate
    private Instant updatedAt;
}
